### ISSUE TITLE
[Global] Page title identical throughout

### ISSUE THREAD

**Issue by nicklepine** [Shopify/dawn#77]
## Issue summary

The page `title` content is the same for each page throughout: "OS 2.0." The `title` is the first bit of content conveyed to screen reader users. As a result, screen reader users rely on unique page `title` content to immediately confirm the current position within the site.

In addition, context-aware `title` content is also required to front-load critical details of the current state. Without this context, users may be unaware of the current state of a page and need to search around the page content to understand the current context.

<details>
  <summary>Screenshots</summary>
<img width="1356" alt="Screenshot of Dawn theme product page. Browser tab content of OS 2.0 is highlighted. Web inspector reveals the title element which is also highlighted." src="https://user-images.githubusercontent.com/1392632/120536791-80f3fa00-c3b2-11eb-83a3-4537801462b9.png">

</details>

#### HTML

```html
<title>OS 2.0</title>
```

## Steps to reproduce

1. Go to a product page, using Dawn with any supported browser
2. Review the `title` content in the browser tab
3. With an active screen reader, navigate through the main navigation and load a new page

### Expected

- `title` content to reflect the current page context.

### Actual

- `title` content is the same throughout.

## Recommendation

Adjust the `title` content to be based on the current page content and state.

Include context-aware state to front-load information right away for screen reader users.

#### HTML

```html
<!-- Homepage -->
<title>OS 2.0</title>

<!-- Product page -->
<title>{Product title} – OS 2.0</title>

<!-- Search results page with results found -->
<title>9 results found for "sandal" – Search results – OS 2.0</title>

<!-- Search results page with no results found -->
<title>No results found for "glasses" – Search results – OS 2.0</title>

<!-- Contact page in error state -->
<title>Error – Contact – OS 2.0</title>
```

<!-- #### Diff

```diff
- line removed
+ line added
``` -->

## Specifications

- **Component:** Global
- **WCAG Principal:** Operable
- **WCAG SC:** [2.4.2 Page Titled](https://www.w3.org/WAI/WCAG21/Understanding/page-titled.html)
- **Severity:** High
- **Effort:** Medium

**Comment by PaulNewton** [comment]
> In addition, context-aware title content is also required to front-load critical details of the current
This also affects bookmarks default text having not context which becomes an organizational problem when you have multiple bookmarks for a site.
 And of course in desktop browsers themselves having multiple tabs|windows open finding the right one quickly becomes a problem when either the titles are all the same or the beginning text of the title is the same with tabs not showing the entire title.

**Comment by PaulNewton** [comment]
Titles are complex overlooked things with many factors in: accessibility, browser UX, navigation, feedback(_errors,search_), SEO, and organization(_bookmarks,tabs,workflows_) . Titles are literally in most cases the first detail a customer may ever encounter about a merchants store before they even arrive at a page.
Structure is needed for something with so many factors that can create endless configurations, _and endless opinions_. 

 With that in mind unless the intent is for <title> in the dawn reference theme to be as reductive as possible and uninformative to developers then an inventory should be done to better define what titles actually **are** in shopify before we can make robust best practices.   
 
Define the concepts, constituent categories and parts along with the construction rules of a title as it applies to a shopify store and the terms we use to refer to these things when designers are creating themes without expecting them to have to have expertise in all of the above disciplines while also having to navigate shopify's architecture.
 _Note that the default logic and text of `page_title` in different contexts is undocumented in liquid reference , such as what text is shown for a search page , or tags that don't exist, etc it's a loooong list of edge cases_.

### What is a title made of?
Example concepts: two general top level categories **Core-contexts** , and **Extended-contexts-hints**(?_meta-hints_?).
Two other specific categories **Settings-context** and **Merchant-contexts**. 
And maybe two mode contexts 1) `seo_title` read by humans and bots, and  )2 `general_title`  private titles only seen by humans (_$900 cart total or Bob your order was shipped!_).
(_when there's better standardized use those please_).
`Delimiters `( -, &, ~) as they are often repeated throughout code but rarely exposed as merchant configurable or localized.
`HTML support` - most text output being put into a <`title`> must be html escaped to prevent serious parsing issues in the <`head`> of the html document.
`Iconography` - ascii, unicode symbols and emoji (🔍,⚠️,🙌,etc) often at the start of a title for fast recognition of a page title or use in product shortlisting or workflows.
`Ordering` - most front facing systems have titles that are read specific to general (left to right) , a default having the storename be at the end of the title becoming more and more specific towards the start of the title, though often the start may have the category of the current page context i.e. "Search - results for.."

### How to evaluate what a title should be made of and exist in the reference theme?
`Legibility and Readibility` - Can a person read the title as a valid sentence?(symbolism or special characters kept to minimum)? Can someone say the title outloud without sounding insane? (_Red shoes contact us closed_ <- preorder contact but brick&mortar  or chat closed for the day)
`Title stub` - the short starting text of a title(~2-3 words) to be considered when figuring out the priority in which to display contexts i.e. "_Search Results for_" vs "_Here are the search results you sear_...." see nngroup https://www.nngroup.com/articles/first-2-words-a-signal-for-scanning/
`Navigation duplication` - situations to be avoided such as any search page results only showing "Search" as the entire title, or several t-shirt collections showing only "T-shirt" on both /womens-tshirt and /mens-tshirt because the text is being stripped. There's also the hard to code situation of storenames being brandnames being collection names being product names or using general terms so the title ends up being [Nike - Nike - Nike - Nike](https://en.wikipedia.org/wiki/Buffalo_buffalo_Buffalo_buffalo_buffalo_buffalo_Buffalo_buffalo)
`Anti-patterns` - example replacing the entire title with only a personalization message "Hi Paul welcome back" and omitting content context. Scrolling marquee or blinking titles.
`Liquid complexity` - Rating of code difficulty to render that context. Some contexts definitely need to be in the dawn reference as they are too complex to expect merchants or designers to discover easily on their own such as the text for a search results pages which has multiple parts **AND** also has translations to account for. _Contrast with something such as building information from metafields which while complex is better suited to a merchant-context and not in a reference theme._
`Bookmarks and workflows` - how useful are the titles contexts when working with many many pages? Customers often build lists of products for comparison and evaluation before making a choice. Can they identify specific pages in the first few words or are they forced to inspect the url of each link lengthening the time of their decision making?


### So what are the Contexts? 🤔
In this example **Core-contexts** are top level places(Index,cart,checkout),  objects(Collections, Products) and their related default title or SEO `page_title`', or symbols(🔍,⚠️,🙌,etc) **before** the default site name. This often reflects information in the url itself i.e a product.handle , subdomain, or the region(i.e USA, AUS,etc). Different from a default context where the title is either hardcoded or always defaults to the page_title with zero logic.
**Extended-contexts-hints** in this example  would be **sub** properties and information of the current core-context, with the URL not necessarily having to contain any aspect of that information. This would also apply to urls that are not indexed such as collections filtered with tags(`collections/collect/tag1+tag2`). Generally generated using liquids `request.page_type` as conditional logic.  So on products that would be something like prepending the product.price, sales status, availability status or a currently selected variant full-name in place of the product.title  , or showing the current SKU.
**Settings-context** are based on theme settings such as products showing the vendor, breadcrumbs, variant deep linking, collection tag display, etc and can be very ambigious as to whether a themes intent aligns with that settings. _An example is if shopify's warehouse theme was a reference if would be helpful to have vendors|brands to be part of the title when it's NOT in the product name._
 **Merchant-contexts** would be specific business rules that probably should be ignored by a reference theme , such as countdown timers, prepending the customers name to every pages, scrolling titles, phone numbers, image galleries, specific SEO marketing strategies(keyword stuffing),  custom error pages(_404 whimsy_), etc . Advanced situations left as end user customizations. **Loosely** includes shopify feature contexts such as local delivery or subscriptions as this may or may not be available to a merchants plan , activated,  or used.

# Liquid bits
Example basic liquid for rendering some of the recommended html with notes on common problems, **before** localization.
_Standalone examples not using page type detection conditionals._
```
<!-- Homepage -->
<title>{{ page_title | default: shop.name}}</title>

<!-- Product page -->
<title>&#123;{{product.title}}&#125; – {{shop.name}}</title>
<!-- note to use curly brackets for html thent he html entities &#123; &#125; should be used for left { and right } curly braces respectively to avoid issues with the liquid parser -->

<!-- Search results page with results found -->
<title>9 results found for "sandal" – Search results – OS 2.0</title>
```
~~`<title> {{ search.results_count }} results found for '{{ search.terms }}' - {{page_title | remove: shop.name }} - {{shop.name}}</title>`~~
Beware a default page_title would be 'Search:  9 results for "sandal" creating duplicate text'; 
and that the `remove `filter used to prevent duplicate words could lead to weirdness for terms that share the same name as the shop itself e.g. nike. _Also there is no ordinal number filter in liquid i.e.. 1 = One, 2 = Two, etc_
Leaving `<title> {{ search.results_count }} {{ search.results_count | pluralize:'result' , 'results' }} found for '{{ search.terms }}' - Search results - {{shop.name}}</title>`
```
<!-- Search results page with no results found -->
<title>No results found for "glasses" – Search results – OS 2.0</title>
<title>{% if search.results_count == 0 %} No {% else %} {{ search.results_count }} {% endif %} results found for '{{ search.terms }}' - Search results - {{shop.name}}</title>
<! -- Note the pluralize filter does not have a zero|empty mode https://shopify.dev/api/liquid/filters/string-filters#pluralize -->

<!-- Contact page in error state -->
<title>Error – Contact – OS 2.0</title>
<title>{% if form.errors %}Error  – {% endif %}{{page_title}} – OS 2.0</title>
<!-- notice how the delimter is inside the conditional  -->
<!-- beware alot of themes use javascript and ajax on forms and are not in the habit of addressing page titles -->
```
_*it's not clear to me yet if new features allow the creation of settings that would let merchants construct the **logic** of page titles using the theme editor blocks ,dynamic sources, etc and never have to touch the liquid_

**Comment by ludoboludo** [comment]
There has been a PR merged already for this that doesn't tackle all use cases but most. We can probably lower the severity and adjust the content of the issue to reflect the newer updates. 

Out of all the the mentioned changes, it seems like the `Error` one is the last one to do/test.
